packages/alfa-selector/src/context.ts

```diff
@@ -0,0 +1,96 @@
+import { Element } from "@siteimprove/alfa-dom";
+import { Map } from "@siteimprove/alfa-map";
+
+export class Context {
+  public static of(state: Iterable<[Element, Context.State]>): Context {
+    return new Context(Map.from(state));
+  }
+
+  private static _empty = new Context(Map.empty());
+
+  public static empty(): Context {
+    return this._empty;
+  }
+
+  private readonly _state: Map<Element, Context.State>;
+
+  private constructor(state: Map<Element, Context.State>) {
+    this._state = state;
+  }
+
+  public hasState(element: Element, state: Context.State): boolean {
+    return this._state.get(element).some((found) => (found & state) !== 0);
+  }
+
+  public getState(element: Element): Context.State {
+    return this._state.get(element).getOr(Context.State.None);
+  }
+
+  public setState(element: Element, state: Context.State): Context {
+    return new Context(this._state.set(element, state));
+  }
```

Comment:
Given the usages we have below, I wonder if we'd need a helper do to the `this.setState(element, this.getState(element) | state)` part. And then just have `hover = foo(Context.State.Hover)`, …

Reply:
Agreed 👍 I'll add an `Context#addState()` method.